cpp/src/command_classes/SwitchBinary.cpp

```diff
@@ -128,24 +158,34 @@ bool SwitchBinary::SetValue
 	Value const& _value
 )
 {
-	if( ValueID::ValueType_Bool == _value.GetID().GetType() )
-	{
-		ValueBool const* value = static_cast<ValueBool const*>(&_value);
+	bool res = false;
+	uint8 instance = _value.GetID().GetInstance();
 
-		Log::Write( LogLevel_Info, GetNodeId(), "SwitchBinary::Set - Setting node %d to %s", GetNodeId(), value->GetValue() ? "On" : "Off" );
-		Msg* msg = new Msg( "SwitchBinaryCmd_Set", GetNodeId(), REQUEST, FUNC_ID_ZW_SEND_DATA, true );
-		msg->SetInstance( this, _value.GetID().GetInstance() );
-		msg->Append( GetNodeId() );
-		msg->Append( 3 );
-		msg->Append( GetCommandClassId() );
-		msg->Append( SwitchBinaryCmd_Set );
-		msg->Append( value->GetValue() ? 0xff : 0x00 );
-		msg->Append( GetDriver()->GetTransmitOptions() );
-		GetDriver()->SendMsg( msg, Driver::MsgQueue_Send );
-		return true;
+	switch( _value.GetID().GetIndex() )
+	{
+		case SwitchBinaryIndex_State:
+		{
+			if( ValueBool* value = static_cast<ValueBool*>( GetValue( _instance, SwitchBinaryIndex_State ) ) )
+			{
+				res = SetState( instance, (static_cast<ValueBool const*>(&_value))->GetValue() );
+				value->Release();
+			}
+			break;
+		}
+		case SwitchBinaryIndex_Duration:
+		{
+			if( ValueByte* value = static_cast<ValueByte*>( GetValue( instance, SwitchBinaryIndex_Duration ) ) )
+			{
+				value->OnValueRefreshed( (static_cast<ValueByte const*>(&_value))->GetValue() );
+				value->Release();
+			}
```

Comment:
Hrm. This will trigger a ValueRefreshed Notification directly from a Application Initiated Change. I dont think we need to do that.... (ie, a Application Setting the ValueID for Duration should know its changed it, we don't need to call back the change). 

(ValueRefreshed Changes are when the device itself has a change to a parameter. in this case, its a local variable, not a device variable).

Reply:
I basically copied this part from https://github.com/OpenZWave/open-zwave/blob/Dev/cpp/src/command_classes/SwitchMultilevel.cpp#L371 to stay consistent with similar code paths. 

I think it's more or less expected of OZW to report back the updated value. At least in ioBroker, there's a difference between parameters sent by the application and parameters accepted by OZW. Or is this handled by node-openzwave-shared anyways?

I'll wait for your reply before changing this part.

Reply:
Ok. I guess we can keep it then... technically it shouldn't happen that way, but I guess we can keep it for now so as to not break existing applications...